### PR TITLE
feat: Initial support for Fares v2 validation.

### PR DESCRIPTION
**Pull request by bdferris-v2**
Per issue #1201, adds validation for files and fields added for Fares v2 base spec changes.

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues

### CODE REVIEW

**Comment by bdferris-v2** [comment]
Initial draft, mostly to see what happens when validating a few existing feeds.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsAnnotationProcessor.java [hunk 1]
```diff
@@ -48,6 +48,10 @@ public class GtfsAnnotationProcessor extends AbstractProcessor {
 
   private final Analyser analyser = new Analyser();
 
+  public GtfsAnnotationProcessor() {
```
Comment (KClough) [inline processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsAnnotationProcessor.java:51]: @bdferris-v2 FYI looks like some debug code may have snuck into this PR

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareTransferRuleSchema.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  String fromLegGroupId();
+
+  @FieldType(FieldTypeEnum.ID)
+  @ForeignKey(table = "fare_leg_rules.txt", field = "leg_group_id")
+  @PrimaryKey(translationRecordIdType = UNSUPPORTED)
+  String toLegGroupId();
+
+  @PrimaryKey(translationRecordIdType = UNSUPPORTED)
+  @Positive
+  int durationLimit();
```
Comment (isabelle-dr) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareTransferRuleSchema.java:44]: Just a question: if we have a negative integer in this field, will this trigger the invalid_integer notice?

Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsFareTransferRuleSchema.java:44]: I can confirm it triggers at `number_out_of_range` validation error.

File: processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsAnnotationProcessor.java (hunk 1 above)
Comment (bdferris-v2) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsAnnotationProcessor.java]: Fixed.   It's kind of a bummer that when the code processor fails, it seems to do so in very obscure ways :(

Comment (KClough) [reply processor/src/main/java/org/mobilitydata/gtfsvalidator/processor/GtfsAnnotationProcessor.java]: 💯 😅

**Comment by bdferris-v2** [comment]
Now that #1248 has been committed and integrated, this PR is ready for official review.

**Review by KClough — APPROVED** [review]
Not an official reviewer, but lgtm.

**Review by isabelle-dr — CHANGES_REQUESTED** [review]
Thanks, @bdferris-v2 🔥
Two things:
1. We are missing documentation for `invalid_fare_transfer_rule_transfer_count` and `fare_transfer_rule_missing_transfer_count`.

2. I would like to include three additional rules in the scope of this PR.
- **`missing_fare_rule_file`**
  - spec mention in [**Dataset files**](https://gtfs.org/schedule/reference/#dataset-files): `fare_rules.txt` is required if fare_attributes.txt is defined.
  - pseudo code: if `fare_attributes.txt `if present and `fare_rules.txt` is absent, trigger the notice.
- **`forbidden_fare_rule_file`**
  - spec mention in [**Dataset files**](https://gtfs.org/schedule/reference/#dataset-files): `fare_rules.txt` is forbidden if `fare_attributes.txt` is not defined.
  - pseudo code: if `fare_attributes.txt `if absent and `fare_rules.txt` is present, trigger the notice.
- **`fare_transfer_rule_with_forbidden_duration_limit`**
  - spec mention in [`fare_transfer_rules.duration_limit_type`](https://gtfs.org/schedule/reference/#fare_transfer_rulestxt):  Forbidden if `fare_transfer_rules.duration_limit` is empty.
  - pseudo code: If [fare_transfer_rules.duration_limit == '' and fare_transfer_rules.duration_limit_type != ''], trigger the notice

I have a copy of the fares v2 spec changes with my comments in [https://bit.ly/faresv2-validator](https://bit.ly/faresv2-validator).

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  /**
+   * A row from GTFS file `fare_transfer_rules.txt` has a defined `transfer_count` with an invalid
+   * value.
+   *
+   * <p>Severity: {@code SeverityLevel.ERROR}
+   */
+  static class InvalidFareTransferRuleTransferCountNotice extends ValidationNotice {
+    private final long csvRowNumber;
+    private final int transfer_count;
```
Comment (maximearmstrong) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java:45]: Could we update this variable to the camel case format for consistency?

**Comment by maximearmstrong** [comment]
> **`fare_transfer_rule_forbidden_duration_limit`**
> * spec mention in [`fare_transfer_rules.duration_limit_type`](https://gtfs.org/schedule/reference/#fare_transfer_rulestxt):  Forbidden if `fare_transfer_rules.duration_limit` is empty.
> * pseudo code: If [`fare_transfer_rules.duration_limit` == '' and `fare_transfer_rules.duration_limit_type` != ''], trigger the notice

@isabelle-dr I think this has been already added with the `FareTransferRuleDurationLimitWithoutTypeNotice` [here](https://github.com/MobilityData/gtfs-validator/pull/1228/files#diff-49a115cb42f4d99e94e0fa80b3086bf1d8952de5b25b845f468f02d391dede16)

**Review by maximearmstrong — CHANGES_REQUESTED** [review]
Great work! I've only one small comment. I think after Isabelle's requested changes are done, we'll be ready to merge. Thanks @bdferris-v2

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java (hunk 3 above)
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java]: Done.

**Comment by bdferris-v2** [comment]
I believe @isabelle-dr was correct that I didn't have a validation check for when `duration_limit_type` was specified but `duration_limit` was not.  I've added that now.

I believe all validation rules have documentation in RULES.md at this point.  Let me know if you see things differently.

As for `fare_rules.txt` and `fare_attributes.txt`, those seem to be clarifications of the Fares V1 spec.  I'm happy to add support for that, but I'd argue for splitting it out into a separate PR to keep things (relatively) focused.  Thoughts?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+   * A row from GTFS file `fare_transfer_rules.txt` has a defined `transfer_count` with an invalid
+   * value.
+   *
+   * <p>Severity: {@code SeverityLevel.ERROR}
+   */
+  static class InvalidFareTransferRuleTransferCountNotice extends ValidationNotice {
+    private final long csvRowNumber;
+    private final int transferCount;
+
+    InvalidFareTransferRuleTransferCountNotice(long csvRowNumber, int transferCount) {
```
Comment (isabelle-dr) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java:47]: @bdferris-v2 I believe this is the rule that checks if `transfer_count` has an integer value outside of [-1;+∞). I don't see this one in the documentation.

**Review by isabelle-dr — COMMENTED** [review]
Hello! Thanks for making this advance 🙏

For the documentation, I believe we are still missing the first rule you have in `FareTransferCountValidator.java` (line 47). It is `invalid_fare_transfer_rule_transfer_count`.

As for the two additional rules I mentioned [here](https://github.com/MobilityData/gtfs-validator/pull/1228#pullrequestreview-1116059614), I see `fare_rules.txt` was introduced in Fares v1, but it went from being optional to conditionally required/forbidden in the PR that added GTFS-Fares v2 base implementation on google/transit ([reference](https://github.com/google/transit/pull/286/files#diff-3ecf0760eb54b4953728042a1e30586705dc2335807be94faae0de5829cd12a1R109)). 
If the scope of this PR is adding validation for the changes introduced in this PR, I would say it's in scope.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java (hunk 4 above)
Comment (bdferris-v2) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareTransferRuleTransferCountValidator.java]: Apologies that I kept missing this.  I renamed this notice to match the other notices in the file and added documentation to RULES.md.

**Review by isabelle-dr — APPROVED** [review]
Looks great!

**Pull request merged**